src/operators/join.rs

```diff
@@ -451,28 +488,20 @@ impl<G, T1> JoinCore<G, T1::Key, T1::Val, T1::R> for Arranged<G,T1>
                 if trace2.is_some() && input1.frontier().is_empty() { trace2 = None; }
                 if let Some(ref mut trace2) = trace2 {
                     trace2.set_logical_compaction(input1.frontier().frontier());
-                    // At this point, if we haven't seen any input batches we should establish a frontier anyhow.
-                    if acknowledged2.is_none() {
-                        acknowledged2 = Some(Antichain::from_elem(<G::Timestamp>::minimum()));
-                    }
-                    if let Some(acknowledged2) = &mut acknowledged2 {
-                        trace2.advance_upper(acknowledged2);
-                        trace2.set_physical_compaction(acknowledged2.borrow());
-                    }
+                    trace2.advance_upper(&mut acknowledged2);
+                    // We validated that `ack2` was at least `distinguish_since()` at start-up.
+                    // It has only advanced since then, so this should be safe to do.
+                    trace2.set_physical_compaction(acknowledged2.borrow());
                 }
 
                 // shut down or advance trace1.
                 if trace1.is_some() && input2.frontier().is_empty() { trace1 = None; }
                 if let Some(ref mut trace1) = trace1 {
                     trace1.set_logical_compaction(input2.frontier().frontier());
-                    // At this point, if we haven't seen any input batches we should establish a frontier anyhow.
-                    if acknowledged1.is_none() {
-                        acknowledged1 = Some(Antichain::from_elem(<G::Timestamp>::minimum()));
-                    }
-                    if let Some(acknowledged1) = &mut acknowledged1 {
-                        trace1.advance_upper(acknowledged1);
-                        trace1.set_physical_compaction(acknowledged1.borrow());
-                    }
+                    trace1.advance_upper(&mut acknowledged1);
```

Comment:
I don't quite understand how `set_logical_compaction`, `advance_upper` and `set_physical_compaction` interact (specifically the first two seem very overlapping?). Haven't read the docs on these so this is a bit of a lazy question - this could be a nice place for a few more comments

Reply:
I'll add some more comments!